packages/alfa-rules/src/common/predicate/has-role.ts

```diff
@@ -1,11 +1,51 @@
 import { Role } from "@siteimprove/alfa-aria";
 import { Element } from "@siteimprove/alfa-dom";
 import { Predicate } from "@siteimprove/alfa-predicate";
+import {hasName} from "./has-name";
 
+const { equals } = Predicate;
+
+export function hasRole(
+  predicate: Predicate<Role>,
+  options?: Role.from.Options
+): Predicate<Element>;
+export function hasRole(
+  name: string,
+  ...rest: Array<string>
+): Predicate<Element>;
 export function hasRole(
-  predicate: Predicate<Role> = () => true,
-  options: Role.from.Options = {}
+  options: Role.from.Options,
+  name: string,
+  ...rest: Array<string>
+): Predicate<Element>;
+export function hasRole(
+  nameOrPredicateOrOptions: string | Predicate<Role> | Role.from.Options,
+  nameOrOptions?: string | Role.from.Options,
+  ...names: Array<string>
 ): Predicate<Element> {
+  let predicate: Predicate<Role>;
+  let options: Role.from.Options;
+
+  if (typeof nameOrPredicateOrOptions === "function") {
+    predicate = nameOrPredicateOrOptions;
+    if (
+      typeof nameOrOptions === "undefined" ||
+      typeof nameOrOptions === "string" // should never happen
+    ) {
+      options = {};
+    } else {
+      options = nameOrOptions;
+    }
+  } else {
+    if (typeof nameOrPredicateOrOptions === "string") {
+      options = {};
+      predicate = hasName(equals(nameOrPredicateOrOptions, nameOrOptions, ...names));
+    } else {
+      options = nameOrPredicateOrOptions;
+      predicate = hasName(equals(nameOrOptions, ...names));
+    }
+  }
+
```

Comment:
We need to factor out the options from this one, otherwise things just get too complex 😮 As such, I'm think that two additional predicates, `hasImplicitRole()` and `hasExplicitRole()`, would be a better way to go about this.

Reply:
Yes, this is clearly too ugly. And I agree that given the options we can have, making specific functions is probably better. It will also make the call easier to read (`hasExplicitRole()` reads better than `hasRole({implicit: false})` which needs to remember that the only other possibility is explicit).